core/src/main/java/org/elasticsearch/threadpool/ScalingExecutorBuilder.java

```diff
@@ -0,0 +1,110 @@
+import org.elasticsearch.node.Node;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Locale;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+
+public class ScalingExecutorBuilder extends ExecutorBuilder<ScalingExecutorBuilder.ScalingExecutorSettings> {
+
+    private final Setting<Integer> coreSetting;
+    private final Setting<Integer> maxSetting;
+    private final Setting<TimeValue> keepAliveSetting;
+
+    public ScalingExecutorBuilder(final String name, final int core, final int max, final TimeValue keepAlive) {
+        this(name, core, max, keepAlive, "thread_pool");
+    }
+
+    public ScalingExecutorBuilder(final String name, final int core, final int max, final TimeValue keepAlive, final String prefix) {
+        super(name);
+        this.coreSetting =
+            Setting.intSetting(settingsKey(prefix, name, "core"), core, Setting.Property.NodeScope);
+        this.maxSetting = Setting.intSetting(settingsKey(prefix, name, "max"), max, Setting.Property.NodeScope);
+        this.keepAliveSetting =
+            Setting.timeSetting(settingsKey(prefix, name, "keep_alive"), keepAlive, Setting.Property.NodeScope);
+    }
+
+    @Override
+    public List<Setting<?>> registerSettings() {
+        return Arrays.asList(coreSetting, maxSetting, keepAliveSetting);
+    }
+
+    @Override
+    public ScalingExecutorSettings settings(Settings settings) {
+        final String nodeName = Node.NODE_NAME_SETTING.get(settings);
+        final int coreThreads = coreSetting.get(settings);
+        final int maxThreads = maxSetting.get(settings);
+        final TimeValue keepAlive = keepAliveSetting.get(settings);
+        return new ScalingExecutorSettings(nodeName, coreThreads, maxThreads, keepAlive);
+    }
+
+    public ThreadPool.ExecutorHolder holder(final ScalingExecutorSettings settings, final ThreadContext threadContext) {
+        TimeValue keepAlive = settings.keepAlive;
+        int core = settings.core;
+        int max = settings.max;
+        final ThreadPool.Info info = new ThreadPool.Info(name(), ThreadPool.ThreadPoolType.SCALING, core, max, keepAlive, null);
+        final ThreadFactory threadFactory = EsExecutors.daemonThreadFactory(EsExecutors.threadName(settings.nodeName, name()));
+        final Executor executor =
+            EsExecutors.newScaling(name(), core, max, keepAlive.millis(), TimeUnit.MILLISECONDS, threadFactory, threadContext);
+        return new ThreadPool.ExecutorHolder(executor, info);
+    }
+
+    @Override
+    public String formatInfo(ThreadPool.Info info) {
+        return String.format(
+            Locale.ROOT,
+            "name [%s], core [%d], max [%d], keep alive [%s]",
+            info.getName(),
+            info.getMin(),
+            info.getMax(),
+            info.getKeepAlive());
+    }
+
+    static class ScalingExecutorSettings extends ExecutorBuilder.ExecutorSettings {
```

Comment:
maybe these settings classes should just accept a `Settings` instance in their ctors? or have a second one?